Sofia Lennbom API - Happy thoughts

solen80a opened the pull request:
Please include your Render link here: https://js-project-api-afon.onrender.com
https://happy-thoughts-happy-mind.netlify.app/

solen80a committed:
Created notes, started with endpoints and installed dependencies.

solen80a committed:
Added query param for messages added today, /messagesfromtoday

solen80a committed:
Connected the root to listEndpoints and moved my html documentation to /documentation.

solen80a committed:
Changed from messages to thoughts

solen80a committed:
Installed and imported mongoose

solen80a committed:
Added Url and Schema

solen80a committed:
Seeded the DB with data from json. Started to transform the code to fetch from DB.

solen80a committed:
Endpoint /thoughts with query params updated to fetch from API

solen80a committed:
Updated endpoint /thoughts/:id to fetch from DB.

solen80a committed:
Some cleaning.

solen80a committed:
Added a POST for new thoughts.

solen80a committed:
Added a DELETE for endpoint /thoughts/:id.

solen80a committed:
Added PATCH for new thougth message and like.

solen80a committed:
Added POSt to increase likes to a thought. Also added message to a like as required and minlength of 5 caracters.

solen80a committed:
Changed the response on my get /thoughts

solen80a committed:
Note update

solen80a committed:
Deployed and connected API and DB. Installed dotenv and created a env file.

solen80a committed:
Added bcrypt. Created models thoughts and users

solen80a committed:
Updated patch

solen80a committed:
Updates to patch

solen80a committed:
Made som comments.

solen80a committed:
Broke out models for user and thought. Created get post for users and authentication.

solen80a committed:
Broke out authanticateuser and created routes for thoughts and user, not broken out yet.

solen80a committed:
Made som cleaning of unused code.

solen80a committed:
Added to include user for edit, delete, post/get thought.

solen80a committed:
More fixes to POST thought to identified user.

solen80a committed:
More authentication fixes.

solen80a committed:
more fixes

solen80a committed:
Cleaning and checking tasks.

solen80a committed:
Update deployed frontend and backend.

ssofiejohansson commented on routes/thoughtRoutes.js:
remove unnecessary file

ssofiejohansson commented on routes/userRoutes.js:
remove unnecessary file

ssofiejohansson commented on utils/postUser.js:
maybe add user input validation so you dont risk empty fields in your database.

if (!email || !password) {
  return res.status(400).json({ success: false, message: "Email and password are required." });
}

ssofiejohansson commented on utils/postUser.js:
```diff
@@ -0,0 +1,26 @@
+    const { email, password } = req.body
+    const salt = bcrypt.genSaltSync()
+    const user = new User({email, password: bcrypt.hashSync(password, salt)})
+    user.save()
```
add: await user.save();

JennieDalgren replied:
don't forget the **await**

ssofiejohansson commented on utils/postUser.js:
```diff
@@ -0,0 +1,26 @@
+
+    })
+  } catch(error){
+    res.status(400).json({
```
Create a visible error message for the user, maybe something like:
res.status(400).json({
  success: false,
  message: "Could not create user",
  error: error.message,
  details: error.errors
});

ssofiejohansson commented on server.js:
Very organised routes! 👏

ssofiejohansson commented on server.js:
```diff
@@ -11,9 +28,229 @@ const app = express()
+})
+
+app.get("/users", async (req, res) => {
+  const { email } = req.params  
```
Use req.query or change to /users/:email

ssofiejohansson commented on models/user.js:
```diff
@@ -0,0 +1,30 @@
+  },
+  password: {
+    type: String,
+    required: true,
+    minlength: 3,
+    maxlength: 100
+  },
```
Maybe also show to the user with a message that there is a password requirement?

ssofiejohansson commented on models/thought.js:
```diff
@@ -0,0 +1,24 @@
+  message: {
+    type: String,
+    required: true,
+    minlength: 5
```
Add a `maxlength` too

ssofiejohansson commented on models/thought.js:
Superclear structure on the models and its easy to follow your code

ssofiejohansson reviewed (commented):
💛 Love the use of the in app pop up notifications!
🙂 The smileyface button to login is is cute but not super clear what it does, maybe add a Login-text too?
🌟 After creating a user, Im navigated to the homepage again, maybe this should be to login page instead so the user dont have to go back.
❌ Add a log out button for better user experience.
🌸 Would be nice to also see what user is currently logged in!
🔙 Add a back button at Login page to Home (pressed it by mistake and couldnt find my way back)
📝 When writing wrong password I still navigate to post a thought, but then get an error trying to post. Easy fix: add error msg when wrong password is submitted ✨
🔥 Show min/max password length for user 

Well done Sofia, you've done a great job. Just a few things I would recommend to change/add to make the experience better for the user 🐝

JennieDalgren commented on models/thought.js:
```diff
@@ -0,0 +1,24 @@
+    type: Date,
+    default: Date.now
+  },
+  user: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: "User",
+    required: true
+  }
```
⭐

JennieDalgren commented on notes.md:
```diff
@@ -0,0 +1,67 @@
+W1:
```
Clear todo list! Love it.

JennieDalgren commented on server.js:
```diff
@@ -11,9 +28,229 @@ const app = express()
+//   }
+//   seedDatabase()
+// }
+
```
IF you dont need this, maybe delete it? Since you have the if statement it is okay to have it "active" if you want to be able to use it with that env variable.

JennieDalgren commented on server.js:
```diff
@@ -11,9 +28,229 @@ const app = express()
+
+})
+
+//Endpoint to show all thoughts
+//Filter liked thoughts, thoughts?liked, 
+//Filter thoughts from today thoughts?thoughtsfromtoday
```
good comments with explanation

JennieDalgren commented on server.js:
```diff
@@ -11,9 +28,229 @@ const app = express()
+  const { newThoughtMessage } = req.body
+
+  try{
+    //const thought = await Thought.findByIdAndUpdate(id, { message: newThoughtMessage }, { new: true , runValidators: true })
```
REmove ?

JennieDalgren reviewed (changes requested):
Great job with this project! And nice implementation of the frontend too.

I'm missing some proper error message when trying to login with non-existing user. It just closes the popup and that's it. As a user I would like to see some message.

Same goes when trying to register as a user. I can't register and I don't get any message about what went wrong.

Take a look at this and ping me when you have worked it out.

solen80a committed:
Fixed user already exist error handling.

solen80a committed:
Removed unused code. Added maxlength to message. Added await to the user save.

solen80a committed:
Added check for password length.

solen80a committed:
removed empty files.

solen80a commented:
Hi, ready for a new review!
I think I have everything working now with the sign in and sing up. Also added more clear messages to the user and added a logout button.

JennieDalgren reviewed (approved):
✅
